### PR TITLE
completions: return non-ok status codes and response headers to clients

### PR DESCRIPTION
**Pull request by bobheadxi**
Right now, all errors from completions providers are provided back to the client with status 500, or worse, status 200 when streaming responses (where we never write any header, so the default 200 is sent on the first write). This change ensures we are propagating non-ok status codes and response headers (important for rate limiting) appropriately back to the client:

1. If status 429, write back 429 to help support rate limit handling in clients
2. If any other non-okay status, write back 503 Service Unavailable to avoid confusion with Sourcegraph status codes while indicating the upstream service encountered a problem

Closes https://github.com/sourcegraph/sourcegraph/issues/52633

## Test plan

Unit tests on error utilities and unit tests on each client to ensure all `CompletionsClient` methods propagate the new typed error.

### CODE REVIEW

**Review by eseliger — COMMENTED** [review]
2 questions from reading this:

- we now forward all error codes 1:1 to the client, right? I could imagine this causing some trouble when we forward a 401 for example, because cody might interpret that as a signout event 🤔 Maybe we should special case 429s
- After reading this, I wonder if we should have some special well typed error for this or so, that also makes clear that this is _not_ coming from the user rate limiter somehow, to not have confusing debugging problems later 

WDYT?

File: enterprise/internal/completions/httpapi/stream.go [hunk 1]
```diff
@@ -32,9 +34,17 @@ func NewChatCompletionsStreamHandler(logger log.Logger, db database.DB) http.Han
...
+		err = cc.Stream(ctx, types.CompletionsFeatureChat, requestParams,
+			func(event types.CompletionResponse) error { return eventWriter.Event("completion", event) })
 		if err != nil {
 			trace.Logger(ctx, logger).Error("error while streaming completions", log.Error(err))
+
+			// Propagate the upstream headers to the client if available.
+			if errNotOK, ok := types.IsErrStatusNotOK(err); ok {
+				errNotOK.WriteResponseHeaders(w)
+			} else {
+				w.WriteHeader(http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/completions/httpapi/stream.go:46]: ahh I see, nice. Looking into that deeper, I wonder if the content-type: text/event-stream might cause issues with parsing on the client?

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/completions/httpapi/stream.go:46]: I think in the case of a non-200 code we should write json or text, I believe the client side is not starting event stream parsing for errored requests 🤔

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/internal/completions/httpapi/stream.go:46]: It looks like the client still somewhat expects the response data to be an event stream: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts?L101:1-118:1

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/completions/httpapi/stream.go:46]: huh okay strange 😀

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/internal/completions/httpapi/stream.go:46]: it's a type of consistency I suppose 🙃

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/internal/completions/httpapi/stream.go:46]: I think you were right here, and this is the root cause of https://github.com/sourcegraph/sourcegraph/issues/53177:

```
[       frontend] ERROR completions.chat httpapi/stream.go:46 error while streaming completions {"trace.family": "session > completions.chat", "error": "Sourcegraph Cody Gateway: unexpected status code 429: you have exceeded the rate limit of 3 requests. Retry after 2023-06-15 18:56:38 +0000 UTC\n"}
[       frontend] 2023/06/14 19:32:59 http: WriteHeader called with both Transfer-Encoding of "chunked" and a Content-Length of 90
[       frontend] ERROR completions.chat httpapi/stream.go:55 error reporting streaming completion error {"trace.family": "session > completions.chat", "error": "http: wrote more than the declared Content-Length"}
```

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/internal/completions/httpapi/stream.go:46]: Ah, the issue turned out to be that we don't _flush_ the headers immediately, but we do add a bunch of headers that makes this behaviour I added invalid. See https://github.com/sourcegraph/sourcegraph/pull/53510

**Pull request merged**